### PR TITLE
Carson/build cxx

### PR DESCRIPTION
**Pull request by ThatsNotVeryCashMoneyOfYou**
This is the PR that adds C++ support and some other things. 

1. This adds the prebuild libc++ libs that we can use to build our stuff and their headers 
2. It changes the build system for dfsan to use our own libcxx, as well as our taint sources 
3. We have a rewrite of polyclang to be in C++ and cleaner, with some better comments 
4. There are improvements to dfsan, specifically some bug fixes found in pdfium, and a new way of dealing with threads
5. There are new docker files for qpdf and pdfium parser support 
6. We have updated the polytrack ABI list with some things so we can build pdfium 

right now this only works in docker containers due to some hard coded paths, i will get to that eventually, i did some hardcoding out of ease because pdfium build system requires things to be in specific places, which is super annoying. you basically have to trick it into thinking we are clang

### CODE REVIEW

File: Dockerfile-pdfium.demo [hunk 1]
```diff
@@ -0,0 +1,41 @@
...
+#For some reason this isnt in our llvm-7 install?
+RUN apt install lld -y
+
+WORKDIR pdfium
+
+#Quick fix
+RUN patch -p1 < /polytracker/pdfium_patch
+
+RUN ln -fs /usr/share/zoneinfo/America/Chicago /etc/localtime
+ENV DEBIAN_FRONTEND=noninteractive
```
Comment (bradlarsen) [inline Dockerfile-pdfium.demo]: We probably only want to set `DEBIAN_FRONTEND` locally in the image-building commands that require that, and not set it persistently as an environment variable using `ENV`.

File: Dockerfile [hunk 2]
```diff
@@ -1,18 +1,20 @@
...
 RUN wget -O - https://apt.llvm.org/llvm-snapshot.gpg.key | apt-key add - && echo "deb http://apt.llvm.org/bionic/ llvm-toolchain-bionic-7 main" | tee -a /etc/apt/sources.list && apt-get -y update && apt-get install -y llvm-7 clang-7
 
+WORKDIR /
+
+RUN git clone https://github.com/RoaringBitmap/CRoaring.git
+
+WORKDIR /CRoaring
+
+RUN mkdir -p build && cd build && cmake .. && make && make install
+
```
Comment (bradlarsen) [inline Dockerfile]: These changes shouldn't be needed, now that the CRoaring and nlohmann-json libraries have been vendored.

File: polytracker/polyclang/polytracker.h [hunk 3]
```diff
@@ -6,13 +6,13 @@
 #define POLYTRACKER_VERSION_REVISION 1
 
 // Set the version note to an empty string if there is no note
-#define POLYTRACKER_SUFFIX     "alpha2"
+#define POLYTRACKER_SUFFIX     "alpha1-SafeDocs-Internal"
```
Comment (ThatsNotVeryCashMoneyOfYou) [inline polytracker/polyclang/polytracker.h]: Fix this

File: Dockerfile-pdfium.demo [hunk 4]
```diff
@@ -0,0 +1,41 @@
...
+RUN gclient config --unmanaged https://pdfium.googlesource.com/pdfium.git
+RUN gclient sync
+
+RUN apt update
+RUN apt install software-properties-common -y
+RUN add-apt-repository "deb http://archive.canonical.com/ $(lsb_release -sc) partner"
+RUN apt update
+RUN apt install sudo -y
+RUN apt install vim -y
+#For some reason this isnt in our llvm-7 install?
```
Comment (bradlarsen) [inline Dockerfile-pdfium.demo]: I think we need to add `apt-get install lld-7` in the base `trailofbits/polytracker` image.

File: polytracker/dfsan/dfsan_rt/dfsan/dfsan.cc [hunk 5]
```diff
@@ -38,17 +38,17 @@
 #include <unordered_set>
 #include <algorithm>
 #include <stack> 
+#include <thread> 
+#include "json.hpp"
```
Comment (bradlarsen) [inline polytracker/dfsan/dfsan_rt/dfsan/dfsan.cc]: this was already present at line 49 and 50 earlier

File: polytracker/dfsan/dfsan_rt/dfsan/dfsan.cc [hunk 6]
```diff
@@ -38,17 +38,17 @@
...
 #include <stdint.h> 
 #include <mutex> 
-#include <ctime>
 //Amalgamated CRoaring lib
 #include "roaring.hh"
 #include "roaring.c" 
-//nlohmann-json
-#include "json.hpp"
 #define MAX_CACHE 100000
+#define POLYTRACKER_ENV_VAR "POLYPATH"
```
Comment (bradlarsen) [inline polytracker/dfsan/dfsan_rt/dfsan/dfsan.cc]: This isn't used.

File: polytracker/dfsan/dfsan_rt/dfsan/dfsan.cc [hunk 7]
```diff
@@ -654,12 +668,20 @@ static void dfsan_init(int argc, char **argv, char **envp) {
...
+	}
+
 	// Register the fini callback to run when the program terminates
 	// successfully or it is killed by the runtime.
 	//
 	// Note: we do this at the very end of initialization, so that if
 	// initialization itself fails for some reason, we don't try to call
 	// `dfsan_fini` from a partially-initialized state.
+	// Register the fini callback to run when the program terminates successfully
+	// or it is killed by the runtime.
```
Comment (bradlarsen) [inline polytracker/dfsan/dfsan_rt/dfsan/dfsan.cc]: This comment shows up a few lines previously, also. Copypasta?

**Comment by bradlarsen** [comment]
It'd also be good to fix the "Build Docker Images" action (should be a one-line change to update a reference to `Dockerfile.demo` to reflect the renaming).

File: Dockerfile-qpdf.demo [hunk 8]
```diff
@@ -0,0 +1,29 @@
+FROM trailofbits/polytracker
+MAINTAINER Carson Harmon <carson.harmon@trailofbits.com>
+
+RUN wget https://www.zlib.net/zlib-1.2.11.tar.gz
+RUN wget http://www.ijg.org/files/jpegsrc.v9c.tar.gz
```
Comment (bradlarsen) [inline Dockerfile-qpdf.demo]: We should grab these via https.

**Pull request merged**